src/metabase/api/common.clj

```diff
@@ -329,6 +335,115 @@
   ([entity id & other-conditions]
    (write-check (apply db/select-one entity :id id other-conditions))))
 
+;;; --------------------------------------------------- STREAMING ----------------------------------------------------
+
+(def ^:private ^:const streaming-response-keep-alive-interval-ms
+  "Interval between sending newline characters to keep Heroku from terminating requests like queries that take a long
+  time to complete."
+  (* 1 1000))
+
+;; Handle ring response maps that contain a core.async chan in the :body key:
+;;
+;; {:status 200
+;;  :body (async/chan)}
+;;
+;; and send each string sent to that queue back to the browser as it arrives
+;; this avoids output buffering in the default stream handling which was not sending
+;; any responses until ~5k characters where in the queue.
+(extend-protocol protocols/StreamableResponseBody
+  clojure.core.async.impl.channels.ManyToManyChannel
+  (write-body-to-stream [output-queue _ ^OutputStream output-stream]
+    (log/debug (u/format-color 'green "starting streaming request"))
+    (with-open [out (io/writer output-stream)]
+      (loop [chunk (async/<!! output-queue)]
+        (cond
+          (char? chunk)
+          (do
+            (try
+              (.write out (str chunk))
+              (.flush out)
+              (catch org.eclipse.jetty.io.EofException e
+                (log/info e (u/format-color 'yellow "connection closed, canceling request"))
+                (async/close! output-queue)
+                (throw e)))
+            (recur (async/<!! output-queue)))
+
+          ;; An error has occurred, let the user know
+          (instance? Exception chunk)
```

Comment:
@metabase/core-developers This is an edge case that I think will happen and we should have a solution. I think it will involve some FE work as well.

The scenario is, what happens when a warehouse is slow to respond and when it does respond, it responds by causing an exception to be thrown. The problem that we have is this code will occasionally send newline characters to the client to check if the client is still there. Sending one of those causes an HTTP response with a 200 status code to be sent. So we have sent a 200 code and maybe a couple of newline characters, now we get a proper exception. I've coded around that in the streaming PR for appdb queries [here](https://github.com/metabase/metabase/pull/6657/files#diff-4a82538ddca88aa9a9b98d5937cde145R299). Where I check to see if I've got one row before proceeding. I can't do that here as we fully realized the results.

My thinking is, we return a data structure that has the error information and the FE needs to treat that 200 response with error payload as if it's a error with a proper 400 code. Calls to POST `/api/dataset` will need to include a check as to whether it has returned an error payload or a query results payload. They could both be JSON with different sets of keys. Would something like that be difficult?

Reply:
@senior Shouldn't be difficult, it will require modifying our `lib/api.js` a bit.